File: app/src/main/java/com/sergiocruz/Matematica/fragment/PrimesTableFragment.java
package com.sergiocruz.Matematica.fragment;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Point;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Display;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.GridView;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.sergiocruz.Matematica.R;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.ArrayList;

import static android.widget.Toast.makeText;
import static com.sergiocruz.Matematica.R.id.card_view_1;
import static java.lang.Integer.parseInt;

/*****
 * Project Matematica
 * Package com.sergiocruz.Matematica.fragment
 * Created by Sergio on 11/11/2016 16:31
 ******/

public class PrimesTableFragment extends Fragment {

    public AsyncTask<Integer, Float, ArrayList<String>> BG_Operation = new LongOperation();
    public ArrayList<String> tableData = null;
    int num_min, num_max, cv_width, height_dip;
    View progressBar;
    GridView history_gridView;
    NumberFormat number_formatter;
    Fragment thisFragment = this;
    Button button;
    Activity mActivity = getActivity();

    public PrimesTableFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // have a menu in this fragment
        setHasOptionsMenu(true);
    }


    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        // Inflate the menu; this adds items to the action bar if it is present.
        inflater.inflate(R.menu.menu_history, menu);

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        if (id == R.id.action_share_history) {
            if (tableData != null) {
                String primes_string = "Tabela de Números Primos entre " +
                        num_min + " e " + num_max + ":\n" +
                        tableData;
                Intent sendIntent = new Intent();
                sendIntent.setAction(Intent.ACTION_SEND);
                sendIntent.putExtra(Intent.EXTRA_TEXT, "Matemática\n" + primes_string);
                sendIntent.setType("text/plain");
                getActivity().startActivity(sendIntent);
            } else {
                Toast thetoast = makeText(getActivity(), "Sem dados para partilhar", Toast.LENGTH_SHORT);
                thetoast.setGravity(Gravity.CENTER, 0, 0);
                thetoast.show();
            }
        }

        if (id == R.id.action_clear_all_history) {
            history_gridView.setAdapter(null);
            tableData = null;
            num_min = num_max = 0;
            EditText min = (EditText) getActivity().findViewById(R.id.min);
            min.setText("");
            EditText max = (EditText) getActivity().findViewById(R.id.max);
            max.setText("");
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final View view = inflater.inflate(R.layout.fragment_primes_table, container, false);

        button = (Button) view.findViewById(R.id.button_gerar_tabela);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                gerar_tabela_primos(view);
            }
        });

        Button btn_clear_min = (Button) view.findViewById(R.id.btn_clear_min);
        btn_clear_min.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                EditText min = (EditText) getActivity().findViewById(R.id.min);
                min.setText("");
            }
        });

        Button btn_clear_max = (Button) view.findViewById(R.id.btn_clear_max);
        btn_clear_max.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                EditText max = (EditText) getActivity().findViewById(R.id.max);
                max.setText("");
            }
        });

        final EditText min_edittext = (EditText) view.findViewById(R.id.min);

        min_edittext.addTextChangedListener(new TextWatcher() {
            Integer num1;
            String oldnum1;

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                oldnum1 = s.toString();
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.toString().equals("")) {
                    return;
                }
                try {
                    // Tentar converter o string para long
                    num1 = parseInt(s.toString());
                } catch (Exception e) {
                    min_edittext.setText(oldnum1);
                    min_edittext.setSelection(min_edittext.getText().length()); //Colocar o cursor no final do texto

                    Toast thetoast = makeText(getActivity(), "Valor mínimo demasiado alto", Toast.LENGTH_LONG);
                    thetoast.setGravity(Gravity.CENTER, 0, 0);
                    thetoast.show();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        final EditText max_edittext = (EditText) view.findViewById(R.id.max);

        max_edittext.addTextChangedListener(new TextWatcher() {
            Integer num2;
            String oldnum2;

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                oldnum2 = s.toString();
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.toString().equals("")) {
                    return;
                }
                try {
                    // Tentar converter o string para long
                    num2 = parseInt(s.toString());
                } catch (Exception e) {
                    max_edittext.setText(oldnum2);
                    max_edittext.setSelection(max_edittext.getText().length()); //Colocar o cursor no final do texto

                    Toast thetoast = makeText(getActivity(), "Valor máximo demasiado alto", Toast.LENGTH_LONG);
                    thetoast.setGravity(Gravity.CENTER, 0, 0);
                    thetoast.show();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        return view;
    }

    public void gerar_tabela_primos(View view) {

        EditText min_edittext = (EditText) view.findViewById(R.id.min);
        String min_string = (String) min_edittext.getText().toString().replaceAll("[^\\d]", "");

        EditText max_edittext = (EditText) view.findViewById(R.id.max);
        String max_string = (String) max_edittext.getText().toString().replaceAll("[^\\d]", "");

        if (min_string.equals(null) || min_string.equals("") || min_string == null || max_string.equals(null) || max_string.equals("") || max_string == null) {
            Toast thetoast = makeText(getActivity(), "Preencher os campos do valor mínimo e máximo", Toast.LENGTH_LONG);
            thetoast.setGravity(Gravity.CENTER, 0, 0);
            thetoast.show();
            return;
        }

        try {
            // Tentar converter o string do valor mínimo para Integer 2^31
            num_min = Integer.parseInt(min_string);
            if (num_min < 2) {
                Toast thetoast = makeText(getActivity(), "Menor número primo = 2", Toast.LENGTH_SHORT);
                thetoast.setGravity(Gravity.CENTER, 0, 0);
                thetoast.show();
                min_edittext.setText("2");
                num_min = 2;
            }
        } catch (Exception e) {
            Toast thetoast = makeText(getActivity(), "Valor mínimo demasiado alto", Toast.LENGTH_LONG);
            thetoast.setGravity(Gravity.CENTER, 0, 0);
            thetoast.show();
            return;
        }

        try {
            // Tentar converter o string do valor mínimo para Integer 2^31
            num_max = Integer.parseInt(max_string);
            if (num_max < 2) {
                Toast thetoast = makeText(getActivity(), "Menor número primo = 2", Toast.LENGTH_SHORT);
                thetoast.setGravity(Gravity.CENTER, 0, 0);
                thetoast.show();
                max_edittext.setText("2");
                num_max = 2;
            }
        } catch (Exception e) {
            Toast thetoast = makeText(getActivity(), "Valor máximo demasiado alto", Toast.LENGTH_LONG);
            thetoast.setGravity(Gravity.CENTER, 0, 0);
            thetoast.show();
            return;
        }


        if (num_min > num_max) {
            int swapp = num_min;
            num_min = num_max;
            num_max = swapp;
            min_edittext.setText(String.valueOf(num_min));
            max_edittext.setText(String.valueOf(num_max));
        }

        BG_Operation = new LongOperation().execute(num_min, num_max);


    }

    private ArrayList<String> getPrimes(int num_min, int num_max) {
        ArrayList<String> primes = new ArrayList<>();
        for (int i = num_min; i <= num_max; i++) {
            boolean isPrime = true;
            if (i % 2 == 0) {
                isPrime = false;
            }
            if (isPrime) {
                for (int j = 3; j < i; j = j + 2) {
                    if (i % j == 0) {
                        isPrime = false;
                        break;
                    }
                }
            }

            if (isPrime) {
                primes.add(Integer.toString(i));
            }

        }
        return primes;
    }

    @Override
    public void onStart() {
        super.onStart();  // Always call the superclass method first
        Log.d("Sergio>>>", "onStart: Starting " + "PrimesTableFragment" + " BG_Operation.getStatus() " + BG_Operation.getStatus());
    }

    @Override
    public void onPause() {
        super.onPause();  // Always call the superclass method first
        Log.i("Sergio>>>", "onPause: Pausing " + "PrimesTableFragment");
    }

    @Override
    public void onStop() {
        super.onStop();  // Always call the superclass method first
        Log.e("Sergio>>>", "onStop: Stopping " + "PrimesTableFragment");
    }


    @Override
    public void onDestroy() {
        super.onDestroy();

        if (BG_Operation.getStatus() == AsyncTask.Status.RUNNING) {
            BG_Operation.cancel(true);
            Toast thetoast = Toast.makeText(getActivity(), "Operação cancelada", Toast.LENGTH_SHORT);
            thetoast.setGravity(Gravity.CENTER, 0, 0);
            thetoast.show();
        }
    }


    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
//         Checks the orientation of the screen
//        if (newConfig.orientation == Configuration.ORIENTATION_LANDSCAPE) {
//            Toast.makeText(getActivity(), "landscape", Toast.LENGTH_SHORT).show();
//        } else if (newConfig.orientation == Configuration.ORIENTATION_PORTRAIT){
//            Toast.makeText(getActivity(), "portrait", Toast.LENGTH_SHORT).show();
//        }
        if (tableData != null) {
            Display display = getActivity().getWindowManager().getDefaultDisplay();
            Point size = new Point();
            display.getSize(size);
            int width = size.x;  //int height = size.y;
            int min_num_length = tableData.get(tableData.size() - 1).toString().length();
            int num_length = min_num_length * 54 + 8;
            int num_columns = (int) Math.round(width / num_length);
            history_gridView.setNumColumns(num_columns);
            final float scale = getActivity().getResources().getDisplayMetrics().density;
            int lr_dip = (int) (4 * scale + 0.5f) * 2;
            cv_width = width - lr_dip;
        }
        hideKeyboard();
    }

    public void hideKeyboard() {
        //Hide the keyboard
        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(getActivity().getCurrentFocus().getWindowToken(), 0);
    }

    public class LongOperation extends AsyncTask<Integer, Float, ArrayList<String>> {

        @Override
        public void onPreExecute() {
            button.setClickable(false);
            button.setText("Working...");
            hideKeyboard();
            history_gridView = (GridView) getActivity().findViewById(R.id.history);
            ViewGroup cardView1 = (ViewGroup) getActivity().findViewById(card_view_1);
            cv_width = cardView1.getWidth();
            progressBar = (View) getActivity().findViewById(R.id.progress);
            number_formatter = new DecimalFormat("#0.00");
            float scale = getActivity().getResources().getDisplayMetrics().density;
            height_dip = (int) (4 * scale + 0.5f);
        }

        @Override
        public ArrayList<String> doInBackground(Integer... params) {
            int num_min = params[0];
            int num_max = params[1];
            ArrayList<String> primes = new ArrayList<>();

            if (num_min == 2) {
                primes.add(Integer.toString(2));
            }
            for (int i = num_min; i <= num_max; i++) {
                boolean isPrime = true;
                if (i % 2 == 0) {
                    isPrime = false;
                }
                if (isPrime) {
                    for (int j = 3; j < i; j = j + 2) {
                        if (i % j == 0) {
                            isPrime = false;
                            break;
                        }
                    }
                }
                if (isPrime) {
                    primes.add(Integer.toString(i));
                }
                float percent = (float) ((float) i / (float) num_max);
                number_formatter.format(percent);
                publishProgress(percent);
            }
            return primes;
        }

        @Override
        public void onProgressUpdate(Float... values) {
            if (thisFragment != null && thisFragment.isVisible()) {
                progressBar.setVisibility(View.VISIBLE);
                int progress_width = (int) Math.round(values[0] * cv_width);
                LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(progress_width, height_dip);
                progressBar.setLayoutParams(layoutParams);
            }
        }

        @Override
        protected void onPostExecute(ArrayList<String> result) {
            tableData = result;
            if (thisFragment != null && thisFragment.isVisible() && result.size() > 0) {
                Display display = getActivity().getWindowManager().getDefaultDisplay();
                Point size = new Point();
                display.getSize(size);
                int width = size.x;  //int height = size.y;
                int max_num_length = result.get(result.size() - 1).toString().length();
                int num_length = max_num_length * 54 + 8;
                int num_columns = (int) Math.round(width / num_length);
                history_gridView.setNumColumns(num_columns);
                ArrayAdapter<String> primes_adapter = new ArrayAdapter<String>(getActivity(), R.layout.table_item, R.id.tableItem, result);
                history_gridView.setAdapter(primes_adapter);
                progressBar.setVisibility(View.GONE);
                button.setClickable(true);
                button.setText("Gerar");
            } else if (result.size() == 0) {
                Toast thetoast = Toast.makeText(getActivity(), "Sem números primos no intervalo", Toast.LENGTH_LONG);
                thetoast.setGravity(Gravity.CENTER, 0, 0);
                thetoast.show();
                history_gridView.setAdapter(null);
                progressBar.setVisibility(View.GONE);
                button.setClickable(true);
                button.setText("Gerar");
            }
        }
    }
}

File: MatematicaApp/src/main/java/com/sergiocruz/MatematicaPro/fragment/PrimesTableFragment.kt
package com.sergiocruz.MatematicaPro.fragment

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.content.res.Configuration
import android.graphics.Point
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.LinearLayout
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.constraintlayout.widget.ConstraintSet
import androidx.recyclerview.widget.GridLayoutManager
import com.sergiocruz.MatematicaPro.BuildConfig
import com.sergiocruz.MatematicaPro.R
import com.sergiocruz.MatematicaPro.activity.AboutActivity
import com.sergiocruz.MatematicaPro.activity.SettingsActivity
import com.sergiocruz.MatematicaPro.adapter.TableAdapter
import com.sergiocruz.MatematicaPro.helper.*
import com.sergiocruz.MatematicaPro.helper.InfoLevel.ERROR
import com.sergiocruz.MatematicaPro.helper.InfoLevel.WARNING
import com.sergiocruz.MatematicaPro.helper.MenuHelper.checkPermissionsWithCallback
import com.sergiocruz.MatematicaPro.helper.MenuHelper.openFolderSnackBar
import com.sergiocruz.MatematicaPro.helper.MenuHelper.saveViewToImage
import kotlinx.android.synthetic.main.fragment_primes_table.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.math.BigInteger
import java.text.DecimalFormat

/*****
 * Project Matematica
 * Package com.sergiocruz.Matematica.fragment
 * Created by Sergio on 11/11/2016 16:31
 */

class PrimesTableFragment : BaseFragment(), OnCancelBackgroundTask, OnEditorActions {

    companion object {
        // isProbablePrime function returns a prime with probability = 1 - (1/2)^certainty
        private const val certainty = 100
    }

    private var asyncTask: AsyncTask<Long, Float, ResultWrapper> = LongOperation()
    private var heightDp: Int = 0
    private var minValue: Long? = 0L
    private var maxValue: Long? = 50L
    private var bruteForceMode: Boolean = true
    private lateinit var tableAdapter: TableAdapter
    private lateinit var layoutManager: GridLayoutManager
    private lateinit var progressBarParams: ConstraintLayout.LayoutParams
    private lateinit var tableData: ResultWrapper

    override fun getHelpTextId(): Int? = null

    override fun getHelpMenuTitleId(): Int? = null

    override fun getHistoryLayout(): LinearLayout? = null

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        getBasePreferences()
        getSharedPreferences()
        writeCalcMode()
    }

    private fun getSharedPreferences() {
        bruteForceMode = sharedPrefs.getBoolean(
            getString(R.string.pref_key_brute_force),
            resources.getBoolean(R.bool.pref_default_brute_force)
        )
    }

    override fun getLayoutIdForFragment() = R.layout.fragment_primes_table

    override fun loadOptionsMenus() = listOf(R.menu.menu_primes_table, R.menu.menu_sub_main)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        getSharedPreferences()
        writeCalcMode()
        heightDp = (4 * scale + 0.5f).toInt()

        switchPrimos.setOnCheckedChangeListener { _, isChecked ->
            if (::tableAdapter.isInitialized)
                tableAdapter.reloadAdapter(isChecked)
        }

        cancelButton.setOnClickListener {
            if (asyncTask.status == AsyncTask.Status.RUNNING || status == OperationStatus.Running) {
                displayCancelDialogBox(context!!, this)
            }
        }
        createTableBtn.setOnClickListener { makePrimesTable() }
        btn_clear_min.setOnClickListener { min_pt.setText("") }
        btn_clear_max.setOnClickListener { max_pt.setText("") }
        min_pt.watchThis(this)
        max_pt.watchThis(this)
    }

    private fun writeCalcMode() {
        calcMode.setText(
            if (bruteForceMode)
                R.string.pref_title_brute_force
            else
                R.string.pref_title_probabilistic
        )
    }

    override fun onActionDone() = makePrimesTable()

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        val childCount = historyGridRecyclerView.childCount
        if (id == R.id.action_save_image_pt) {
            if (childCount > 0) {
                checkPermissionsWithCallback(activity as Activity) {
                    val imgPath = saveViewToImage(historyGridRecyclerView, 0, true)
                    if (imgPath != null) {
                        openFolderSnackBar(activity!!, getString(R.string.image_saved))
                    } else {
                        showCustomToast(this.context, getString(R.string.errorsavingimg), ERROR)
                    }
                }
            } else {
                showCustomToast(context, getString(R.string.empty_table), WARNING)
            }
        }
        if (id == R.id.action_share_history_image_pt) {
            if (childCount > 0) {
                checkPermissionsWithCallback(activity as Activity) {
                    val fileUris = ArrayList<Uri>(1)
                    val imgPath = saveViewToImage(historyGridRecyclerView, 0, true)
                    if (imgPath != null) {
                        fileUris.add(Uri.parse(imgPath))
                        val sendIntent = Intent()
                        sendIntent.action = Intent.ACTION_SEND_MULTIPLE
                        sendIntent.putExtra(
                            Intent.EXTRA_TEXT, getString(R.string.app_long_description) +
                                    BuildConfig.VERSION_NAME + "\n"
                        )
                        sendIntent.type = "image/jpeg"
                        sendIntent.putParcelableArrayListExtra(Intent.EXTRA_STREAM, fileUris)
                        startActivity(
                            Intent.createChooser(
                                sendIntent,
                                resources.getString(R.string.app_name)
                            )
                        )

                    } else {
                        showCustomToast(context, getString(R.string.errorsavingimg), ERROR)
                    }
                }
            } else {
                showCustomToast(context, getString(R.string.empty_table), WARNING)
            }
        }
        // Partilhar tabela como texto
        if (id == R.id.action_share_history) {
            if (tableData?.primesTable?.isNotEmpty()) {
                val primesString =
                    "${getString(R.string.table_between)} $minValue ${getString(R.string.and)} $maxValue:\n${tableData.primesTable}"
                val sendIntent = Intent()
                sendIntent.action = Intent.ACTION_SEND
                sendIntent.putExtra(
                    Intent.EXTRA_TEXT, getString(R.string.app_long_description) +
                            BuildConfig.VERSION_NAME + "\n" + primesString
                )
                sendIntent.type = "text/plain"
                startActivity(sendIntent)
            } else {
                showCustomToast(context, getString(R.string.no_data))
            }
        }

        if (id == R.id.action_clear_all_history) {
            historyGridRecyclerView.adapter = null
            min_pt.text = Editable.Factory().newEditable("1")
            max_pt.text = Editable.Factory().newEditable("50")
            numPrimesTextView.visibility = GONE
            performanceTextView.visibility = GONE
        }
        if (id == R.id.action_about) {
            startActivity(Intent(context, AboutActivity::class.java))
        }
        if (id == R.id.action_settings) {
            startActivity(Intent(context, SettingsActivity::class.java))
        }
        return true
    }

    override fun onOperationCanceled(canceled: Boolean) {
        if (cancelAsyncTask(asyncTask, context)) resetButtons()
        if (status == OperationStatus.Running) {
            status = OperationStatus.Canceled
        }
    }

    private fun makePrimesTable() {
        hideKeyboard(activity)
        val minString = min_pt.text.toString().replace("[^\\d]".toRegex(), "")
        val maxString = max_pt.text.toString().replace("[^\\d]".toRegex(), "")

        if (TextUtils.isEmpty(minString)) {
            min_pt.apply {
                requestFocus()
                error = getString(R.string.fill_this)
                postDelayed({ error = null }, clearErrorDelayMillis)
            }
            showKeyboard(activity)
            return
        }
        if (TextUtils.isEmpty(maxString)) {
            max_pt.apply {
                requestFocus()
                error = getString(R.string.fill_this)
                postDelayed({ error = null }, clearErrorDelayMillis)
            }
            showKeyboard(activity)
            return
        }

        minValue = minString.toLongOrNull(10)
        if (minValue == null) {
            min_pt.apply {
                requestFocus()
                error = getString(R.string.numero_alto)
                postDelayed({ error = null }, clearErrorDelayMillis)
            }
            return
        } else if (minValue!! < 1) {
            min_pt.setText("1")
            minValue = 1
        }

        maxValue = maxString.toLongOrNull(10)
        if (maxValue == null) {
            max_pt.apply {
                requestFocus()
                error = getString(R.string.numero_alto)
                postDelayed({ error = null }, clearErrorDelayMillis)
            }
            return
        } else if (maxValue!! < 1) {
            max_pt.setText("1")
            maxValue = 1
        }

        if (minValue!! > maxValue!!) {
            val swap = minValue
            minValue = maxValue
            maxValue = swap
            min_pt.setText(minValue.toString())
            max_pt.setText(maxValue.toString())
            makePrimesTable()
            return
        }

        lockButtons()
        if (bruteForceMode) {
            setUpAdapter(maxValue!!)
            asyncTask = LongOperation().execute(minValue, maxValue)
        } else {
            status = OperationStatus.Running
            // launch coroutine in the Default thread
            GlobalScope.launch(Dispatchers.Default) {
                val startTime = System.currentTimeMillis()
                val result: ResultWrapper = probabilisticMode(minValue!!, maxValue!!)
                withContext(Dispatchers.Main) {
                    showPerformance(startTime)
                    setUpAdapter(maxValue!!)
                    tableAdapter.swap(result.fullTable, result.primesTable, switchPrimos.isChecked)
                    numPrimesTextView.visibility = VISIBLE
                    numPrimesTextView.text =
                        "${getString(R.string.cardinal_primos)} ${result.primesTable.size}"
                    resetButtons()
                }
                tableData = result
                status = OperationStatus.Done
            }
        }
    }

    /** Data class to wrap the result of the async calculation
     * containing primesTable and fullTable */
    data class ResultWrapper(
        val fullTable: MutableMap<Int, Pair<String, Boolean>>,
        val primesTable: MutableMap<Int, Pair<String, Boolean>>
    )

    private suspend fun probabilisticMode(minValue: Long, maxValue: Long): ResultWrapper {
        val fullTable = mutableMapOf<Int, Pair<String, Boolean>>()
        val primesOnlyTable = mutableMapOf<Int, Pair<String, Boolean>>()
        var indexPrimes = 0
        var progress: Float
        var lastProgress = 0.0f
        for ((index, i) in (minValue..maxValue).withIndex()) {
            fullTable[index] = Pair(i.toString(), false)
            val currentVal: BigInteger = BigInteger.valueOf(i)
            if (currentVal.isProbablePrime(certainty)) {
                primesOnlyTable[indexPrimes] = Pair(i.toString(), true)
                indexPrimes++
                fullTable[index] = Pair(i.toString(), true)
            }

            if (status is OperationStatus.Canceled) {
                status = OperationStatus.Idle
                return ResultWrapper(fullTable, primesOnlyTable)
            }

            progress = i.toFloat() / maxValue.toFloat()
            if (progress - lastProgress > 0.05) {
                updateProgress(progress)
                lastProgress = progress
            }
        }
        return ResultWrapper(fullTable, primesOnlyTable)
    }

    private suspend fun updateProgress(progress: Float) {
        if (this@PrimesTableFragment.isVisible) {
            withContext(Dispatchers.Main) {
                progressBarParams.width = Math.round(progress * cardViewMain.width)
                progressBar.layoutParams = progressBarParams
            }
        }
    }


    private var status: OperationStatus = OperationStatus.Idle

    /**status of the background calculations */
    sealed class OperationStatus {
        object Running : OperationStatus()
        object Canceled : OperationStatus()
        object Done : OperationStatus()
        object Idle : OperationStatus()
    }

    /** Slower method even though it doesn't check all numbers.*/
    private fun probabilisticMode2(minValue: Long, maxValue: Long): ResultWrapper {

        val fullTable = mutableMapOf<Int, Pair<String, Boolean>>()
        val primesOnlyTable = mutableMapOf<Int, Pair<String, Boolean>>()

        var index = 0
        var indexPrimes = 0
        var tracker = minValue
        var nextTracker = tracker

        if (BigInteger.valueOf(minValue).isProbablePrime(certainty)) {
            fullTable[index] = Pair(minValue.toString(), true)
            index++
            primesOnlyTable[indexPrimes] = Pair(minValue.toString(), true)
            indexPrimes++
        }
        val start = BigInteger.valueOf(minValue).nextProbablePrime().toLong()
        for (i in (minValue + index)..(start - 1)) {
            fullTable[index] = Pair(i.toString(), false)
            index++
        }

        while (tracker <= maxValue) {
            tracker = BigInteger.valueOf(tracker).nextProbablePrime().toLong()
            if (tracker >= maxValue) {
                for (i in nextTracker..maxValue) {
                    fullTable[index] = Pair(i.toString(), false)
                    index++
                }
                break
            }

            fullTable[index] = Pair(tracker.toString(), true)
            index++
            primesOnlyTable[indexPrimes] = Pair(tracker.toString(), true)
            indexPrimes++
            nextTracker = BigInteger.valueOf(tracker).nextProbablePrime().toLong()
            if (nextTracker >= maxValue) {
                for (i in (tracker + 1)..maxValue) {
                    fullTable[index] = Pair(i.toString(), false)
                    index++
                }
                break
            }
            for (i in (tracker + 1)..(nextTracker - 1)) {
                fullTable[index] = Pair(i.toString(), false)
                index++
            }
            fullTable[index] = Pair(nextTracker.toString(), true)
            primesOnlyTable[indexPrimes] = Pair(nextTracker.toString(), true)

            if (status == OperationStatus.Canceled) {
                status = OperationStatus.Idle
                return ResultWrapper(fullTable, primesOnlyTable)
            }
        }
        return ResultWrapper(fullTable, primesOnlyTable)
    }

    private fun setUpAdapter(maxValue: Long) {
        tableAdapter = TableAdapter()
        tableAdapter.setHasStableIds(true)
        historyGridRecyclerView.adapter = tableAdapter
        historyGridRecyclerView.setHasFixedSize(true)
        layoutManager = GridLayoutManager(context, getNumColumns(maxValue))
        historyGridRecyclerView.layoutManager = layoutManager
    }

    private fun getNumColumns(maxValue: Long): Int {
        val display = activity?.windowManager?.defaultDisplay
        val size = Point()
        display?.getSize(size)
        val width = size.x  //int height = size.y;
        val numMaxLength = maxValue.toString().length
        val scale = resources.displayMetrics.density
        val numLength = numMaxLength * (18 * scale + 0.5f).toInt() + 8
        return Math.round((width / numLength).toFloat())
    }

    override fun onDestroy() {
        super.onDestroy()
        cancelAsyncTask(asyncTask, context)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        //Quando altera a orientação do ecrã
        if (tableData?.fullTable?.isNotEmpty()) {
            layoutManager.spanCount = getNumColumns(maxValue!!)
        }
        hideKeyboard(activity)
    }

    private fun lockButtons() {
        createTableBtn.isClickable = false
        createTableBtn.text = getString(R.string.working)
        cancelButton.visibility = VISIBLE
        progressBarParams = progressBar.layoutParams as ConstraintLayout.LayoutParams
        progressBarParams.width = 1
        progressBar.layoutParams = progressBarParams
        progressBar.visibility = VISIBLE
        hideKeyboard(activity as Activity)
        numPrimesTextView.visibility = GONE
        performanceTextView.visibility = GONE
    }

    private fun resetButtons() {
        progressBar.visibility = GONE
        cancelButton.visibility = GONE
        createTableBtn.isClickable = true
        createTableBtn.setText(R.string.gerar)
    }

    inner class LongOperation : AsyncTask<Long, Float, ResultWrapper>() {

        private var startTime: Long = System.currentTimeMillis()

        override fun doInBackground(vararg params: Long?): ResultWrapper {
            var min = params[0] as Long
            val max = params[1] as Long

            var index = 0
            var indexPrimes = 0
            val fullTable = mutableMapOf<Int, Pair<String, Boolean>>()
            val primesTable = mutableMapOf<Int, Pair<String, Boolean>>()

            var progress: Float
            var oldProgress = 0.0f
            if (min == 1L) {
                min = 2L
                fullTable[index] = Pair("1", false)
                index++
            }

            if (min == 2L) {
                fullTable[index] = Pair("2", true)
                primesTable[indexPrimes] = Pair("2", true)
                index++
                indexPrimes++
                min = 3
            }

            for (i in min..max) {
                var isPrime = true
                if (i % 2 == 0L) isPrime = false
                if (isPrime) {
                    var j = 3
                    while (j < i) {
                        if (i % j == 0L) {
                            isPrime = false
                            break
                        }
                        j += 2
                    }
                }
                if (isPrime) {
                    fullTable[index] = Pair(i.toString(), true)
                    primesTable[indexPrimes] = Pair(i.toString(), true)
                    index++
                    indexPrimes++
                } else {
                    fullTable[index] = Pair(i.toString(), false)
                    index++
                }

                progress = i.toFloat() / max.toFloat()
                if (progress - oldProgress > 0.05) {
                    publishProgress(progress)
                    oldProgress = progress
                }
                if (isCancelled) break
            }
            return ResultWrapper(fullTable, primesTable)
        }

        override fun onProgressUpdate(vararg values: Float?) {
            if (this@PrimesTableFragment.isVisible) {
                val progress: Float = values[0] ?: 0.0f
                progressBarParams.width = Math.round(progress * cardViewMain.width).toInt()
                progressBar.layoutParams = progressBarParams
            }
        }

        override fun onPostExecute(result: ResultWrapper) {

            if (this@PrimesTableFragment.isVisible) {

                tableAdapter.swap(result.fullTable, result.primesTable, switchPrimos.isChecked)

                numPrimesTextView.visibility = VISIBLE
                numPrimesTextView.text =
                    "${getString(R.string.cardinal_primos)} ${result.primesTable.size}"
                showPerformance(startTime)
                resetButtons()
                tableData = result
            }
        }

        override fun onCancelled(parcial: ResultWrapper) { //resultado parcial obtido após cancelar AsyncTask
            super.onCancelled(parcial)

            if (this@PrimesTableFragment.isVisible && parcial.fullTable.isNotEmpty()) {

                tableAdapter.swap(parcial.fullTable, parcial.primesTable, switchPrimos.isChecked)

                numPrimesTextView.visibility = VISIBLE
                numPrimesTextView.text =
                    "${getString(R.string.cardinal_primos)} (${parcial.primesTable.size})"
                showPerformance(startTime)
                resetButtons()
            } else if (parcial.primesTable.isEmpty()) {
                showCustomToast(context, getString(R.string.canceled_noprimes))
                historyGridRecyclerView.adapter = null
                resetButtons()
            }
            tableData = parcial
        }

    }

    private fun showPerformance(startTime: Long) {
        if (shouldShowPerformance) {
            ConstraintSet().apply {
                clone(cardViewMain)
                connect(
                    R.id.performanceTextView,
                    ConstraintSet.TOP,
                    R.id.numPrimesTextView,
                    ConstraintSet.TOP,
                    0
                )
                connect(
                    R.id.performanceTextView,
                    ConstraintSet.BOTTOM,
                    R.id.numPrimesTextView,
                    ConstraintSet.BOTTOM,
                    0
                )
                setVisibility(R.id.performanceTextView, VISIBLE)
                applyTo(cardViewMain)
            }

            val elapsed =
                " " + DecimalFormat("#.###").format((System.currentTimeMillis() - startTime) / 1000.0) + "s"
            performanceTextView.text = getString(R.string.performance) + " " + elapsed
        } else {
            performanceTextView.visibility = GONE
        }
    }

}
